windows/sound.cpp

```diff
@@ -60,6 +60,13 @@ QString CSound::LoadAndInitializeDriver ( QString strDriverName,
         return tr ( "The current selected audio device is no longer present in the system." );
     }
 
+    // save number of channels from last driver (if they are set); this is done to check if we need to reset channel mapping

```

Comment:
The `if` statement doesn't implement this comment.  The `if` says "save the number of channels if I haven't already saved the number of channels before".  What is the actual intent here?

Reply:
The intent here was to check if we're not on the first start. That's right.

Channel mapping might be reset on the first start if we don't do this (and that might be the bug we want to solve). INVALID_INDEX will never be the correct number of channels? 

I think the comment was added at the wrong position and might be outdated from one of my first tests.

Reply:
Just writing down my thoughts...

So at launch, they're INVALID_INDEX.  `strCurDevName` appears to be playing a similar role to the `lNum*ChanPrev` pair here - hold the "old" value until we've established we're okay, then check what happened (gets initialised in both Windows CSound and base CSoundBase).

If 
- we acquire the driver successfully the first time through (first `CheckDeviceCapabilities();`) - `strStat.isEmpty()` is true
  and (
  - the driver name changed (which is should do on success the first time in)
    or
  - the number of ins or outs changed (which they should do on success the first time in)

  )

then we call `ResetChannelMapping();` and store the values (ins, outs, driver name).

Next time we acquire "some driver" (maybe the same one, maybe different), we save the "old" values num chan values here before they're overwritten by `CheckDeviceCapabilities();`.  That's different from the driver name because it doesn't get overwritten.

So then, we can check whether any value _changed_ and only call `ResetChannelMapping();` if it did.

----

Can you make the comment say "need to save these (but not the driver name), as `CheckDeviceCapabilities();` overwrites them" or words to that effect.  (The reason can be seen in the later code, I don't think it needs to be commented here.  Mentioning the driver name will make the association with the later `if` clear enough.)

I'm still not at all sure the `if` is needed.  Maybe `lNumInChan` and `lNumOutChan` should get set to `INVALID_INDEX` in ` windows/sound.h` instead?

Reply:
I think you're right with this analysis.

A few questions we might want to ask:

What happens and what should happen on the first invocation of LoadAndInitializeDriver()? Do we need to reset channel mapping or not? Did reloading channel mapping here result in the mentioned bug?


We set the channel mapping here for the first time, I think:

https://github.com/jamulussoftware/jamulus/blob/b398f7aaf8909ffbcb13c1970d0ae2960939efed/windows/sound.cpp#L590

What should the value of lNum*ChanPrev be on the first start?

Reply:
If `lNum*Chan` are `INVALID_INDEX` first time through, then leaving the `if` out would mean `lNum*ChanPrev` get set to `INVALID_INDEX` here (regardless of what they were initialised to, so could default to 0.

If the driver is successfully acquired, `lNum*Chan` get the new numbers.  As that means a change happened, `ResetChannelMapping();` gets called (still first time through).

We then _don't need to store_ `lNum*Chan` inside the `ResetChannelMapping();` `if` block -- as they'll get stored here next time through.  ...  Assuming this logic is working of course...

Next time through, we store `lNum*Chan` here.  Then, if the driver name and `lNum*Chan` don't change, `ResetChannelMapping();` isn't called, otherwise it is (and the driver name is updated).

With this logic, `lNum*ChanPrev` seem to be _local variables_ in usage, too -- they don't need to persist across calls, just across the acquiring of the driver capabilities.

Reply:
> If lNum*Chan are INVALID_INDEX first time through, then leaving the if out would mean lNum*ChanPrev get set to INVALID_INDEX here (regardless of what they were initialised to, so could default to 0.

Yes.

> If the driver is successfully acquired, lNum*Chan get the new numbers. As that means a change happened, ResetChannelMapping(); gets called (still first time through).

Are you sure that this is what we want? Currently we don't do that.

Reply:
Hmm. Maybe was a bit confused... The comment should be clear now.

Reply:
@pljones will check your logic but not now.

Reply:
Ok. At least my bug doesn't seem to occur with this logic @pgScorpio @JoeSouthern

Could you please check if this binary works for you: https://github.com/jamulussoftware/jamulus/suites/2375362007/artifacts/50403146

Reply:
> Ok. At least my bug doesn't seem to occur with this logic @pgScorpio @JoeSouthern
> 
> Could you please check if this binary works for you: https://github.com/jamulussoftware/jamulus/suites/2375362007/artifacts/50403146

Finally got some time to do a quick test and this version seems OK too, At least concerning the channel mapping reset.
I still have some comments on other issues though:
Window title:  Server info is moved to the end of the title (Great!), but this only takes effect when using hostname, without using hostname on the commandline the server info is still in front of "Jamulus".